### PR TITLE
twitch-emotes

### PR DESCRIPTION
**Pull request by Gantonator**
Allow twitch emotes in the !alert and !hotseat messages, can be disabled by setting twitchEmotes to false.

### CODE REVIEW

File: twitchpopups.htm [hunk 1]
```diff
@@ -149,9 +157,33 @@ <h1 id="popuptext"></h1>
         $("#popuptext").animate({"opacity":0, "margin-left":"50px"}, 700);
     }
     
+    function formatEmotes(message, emotes) {
+		  //parse the message for html and remove any tags
+      var newMessage = $($.parseHTML(message)).text().toUpperCase().split("");
```
Comment (Rcjuk) [inline twitchpopups.htm:162]: let ;)

File: twitchpopups.htm [hunk 2]
```diff
@@ -149,9 +157,33 @@ <h1 id="popuptext"></h1>
...
     }
     
+    function formatEmotes(message, emotes) {
+		  //parse the message for html and remove any tags
+      var newMessage = $($.parseHTML(message)).text().toUpperCase().split("");
+
+			//replace any twitch emotes in the message with img tags for those emotes
+      if(twitchEmotes){
+        for (var emoteIndex in emotes) {
+          var emote = emotes[emoteIndex];
```
Comment (Rcjuk) [inline twitchpopups.htm:167]: const

File: twitchpopups.htm [hunk 3]
```diff
@@ -149,9 +157,33 @@ <h1 id="popuptext"></h1>
         $("#popuptext").animate({"opacity":0, "margin-left":"50px"}, 700);
     }
     
+    function formatEmotes(message, emotes) {
+		  //parse the message for html and remove any tags
+      var newMessage = $($.parseHTML(message)).text().toUpperCase().split("");
+
+			//replace any twitch emotes in the message with img tags for those emotes
+      if(twitchEmotes){
+        for (var emoteIndex in emotes) {
```
Comment (Rcjuk) [inline twitchpopups.htm:166]: If emotes is an array

```
emotes.forEach((e, i) => {

});
```

File: twitchpopups.htm [hunk 4]
```diff
@@ -149,9 +157,33 @@ <h1 id="popuptext"></h1>
...
+          var emote = emotes[emoteIndex];
+          for (var charIndexes in emote) {
+            var emoteIndexes = emote[charIndexes];
+            if (typeof emoteIndexes == "string") {
+              emoteIndexes = emoteIndexes.split("-");
+              emoteIndexes = [parseInt(emoteIndexes[0]), parseInt(emoteIndexes[1])];
+              for (var i = emoteIndexes[0]; i <= emoteIndexes[1]; ++i) {
+                newMessage[i] = "";
+              }
+              newMessage[emoteIndexes[0]] = "<img class=\"emoticon\" src=\"https://static-cdn.jtvnw.net/emoticons/v1/" + emoteIndex + "/3.0\">";
```
Comment (Rcjuk) [inline twitchpopups.htm:176]: If you use template literals you wont have to escape anything
```
 newMessage[emoteIndexes[0]] = `<img class="emoticon" src="https://static-cdn.jtvnw.net/emoticons/v1/${emoteIndex}/3.0"/>`;
```

File: twitchpopups.htm [hunk 5]
```diff
@@ -149,9 +157,33 @@ <h1 id="popuptext"></h1>
         $("#popuptext").animate({"opacity":0, "margin-left":"50px"}, 700);
     }
     
+    function formatEmotes(message, emotes) {
```
Comment (Rcjuk) [inline twitchpopups.htm:160]: ```const formatEmotes = (message, emotes) => {```

File: twitchpopups.htm (hunk 1 above)
Comment (Gantonator) [reply twitchpopups.htm:162]: oh yeah i did mean to do that

File: twitchpopups.htm (hunk 5 above)
Comment (Gantonator) [reply twitchpopups.htm]: nice nice done done

File: twitchpopups.htm (hunk 4 above)
Comment (Gantonator) [reply twitchpopups.htm]: nice done by d9d19fe

File: twitchpopups.htm (hunk 3 above)
Comment (Gantonator) [reply twitchpopups.htm:166]: It's a hash set or dictionary or something who knows what its called in js but anyway this didn't work so I just left it as it is

File: twitchpopups.htm (hunk 2 above)
Comment (Gantonator) [reply twitchpopups.htm]: done in d9d19fe

File: twitchpopups.htm (hunk 1 above)
Comment (Gantonator) [reply twitchpopups.htm]: done in d9d19fe

File: twitchpopups.htm [hunk 6]
```diff
@@ -92,14 +100,14 @@ <h1 id="popuptext"></h1>
...
-        
+        const formattedMessage = formatEmotes(commandName, context.emotes)
         // If the command is known, let's execute it
 
             if (context.mod || (context["badges-raw"] != null && context["badges-raw"].startsWith("broadcaster"))) {
                 if (commandName.startsWith("!alert ")) {
                     $("#popupbox").show();
                     $("#popupbox").css({"background-color":alertBg}); 
-                    $("#popuptext").text(commandName.substr(7).toUpperCase());
+                    $("#popuptext").html(formattedMessage.substr(7));
```
Comment (cohan) [inline twitchpopups.htm:110]: This was changed early on from .html() to .text() as html was allowing users to say HTML tags in chat, e.g. ``

Just a heads up we'll need to sanitise HTML if this goes through.

Comment (Gantonator) [reply twitchpopups.htm:110]: the new function "formatEmotes()" DOES sanitise the html by using text() first and THEN returns a string containing  tags (which are intentionally not sanitised).

No change necessary.

Comment (cohan) [reply twitchpopups.htm:110]: Ah nice! My apologies :)

**Comment by DaftLimmy** [comment]
I love this, @Gantonator. But it looks like the emotes are a pixel or two too high, can you see it?

Also, does this load in the tiny emotes or the biggest ones? My preference is the 112x112 emotes, obviously, and scale them down if necessary.

**Comment by Gantonator** [comment]
They were a pixel off center vertically.  I agree that this was unacceptable but please don't judge me as harshly as you did the people who initially did the dvd menu for Limmy's show.

I have changed the css so that they are 6px from the top and 6px from the bottom of the box.

This does use 112x112 version of the emotes i.e. the 3.0 version here https://static-cdn.jtvnw.net/emoticons/v1/1936644/3.0 and scales them down.

I did notice that even though the twitch emotes are centered: standard emoji behave like normal text and so aren't aligned in the same way.

image

If you'd prefer for twitch emotes to be aligned just like emoji then the styles to do that would be
`#popuptext img {
    width: 69px;
    margin-bottom: -13px;
}`

image

Yours,
Gantee

**Comment by Gantonator** [comment]
Ok closing this! I'm making a new PR that's a bit nicer and easier to review.

**Pull request closed without merging**